styx-scheduler-service/src/main/java/com/spotify/styx/Scheduler.java

```diff
@@ -251,49 +253,53 @@ private void triggerBackfills(Collection<InstanceState> activeStates) {
             counting()));
 
     backfills.forEach(backfill -> {
-      final BackfillBuilder builder = backfill.builder();
 
       final Optional<Workflow> workflowOpt = workflowCache.workflow(backfill.workflowId());
 
       if (!workflowOpt.isPresent()) {
         LOG.warn("workflow not found for backfill, skipping rest of triggers: {}", backfill);
+        final BackfillBuilder builder = backfill.builder();
         builder.halted(true);
         storeBackfill(builder.build());
         return;
       }
 
-      final int needed = backfill.concurrency() - backfillStates.getOrDefault(backfill.id(), 0L).intValue();
-      if (needed <= 0) {
-        return;
-      }
-
       final Workflow workflow = workflowOpt.get();
 
-      final List<Instant> partitionsRemaining =
-          ParameterUtil.rangeOfInstants(backfill.nextTrigger(), backfill.end(),
-                                        workflow.schedule().partitioning());
-
-      final List<Instant> partitionsNeeded =
-          partitionsRemaining.stream().limit(needed).collect(toList());
-
-      partitionsNeeded.forEach(
-          partition -> {
-            try {
-              triggerListener.event(workflow, Trigger.backfill(backfill.id()), partition);
-            } catch (AlreadyInitializedException e) {
-              LOG.warn("tried to trigger backfill for already active state [{}]: {}",
-                       partition, backfill);
-            }
-          });
-
-      if (partitionsRemaining.size() > partitionsNeeded.size()) {
-        builder.nextTrigger(partitionsRemaining.get(partitionsNeeded.size()));
-      } else {
-        builder.nextTrigger(backfill.end());
-        builder.allTriggered(true);
+      final int remainingCapacity =
+          backfill.concurrency() - backfillStates.getOrDefault(backfill.id(), 0L).intValue();
+
+      Instant partition = backfill.nextTrigger();
+
+      for (int i = 0; i < remainingCapacity && partition.isBefore(backfill.end()); i++) {
+        try {
+          triggerListener
+              .event(workflow, Trigger.backfill(backfill.id()), partition)
+              .toCompletableFuture().get();
```

Comment:
Do we want to have similar comment like https://github.com/spotify/styx/pull/78/files#diff-8681f2f92e0c50ca9a99b3fc6ee3e4a2R105